Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp

```diff
@@ -55,161 +55,81 @@ FileSystemSyncAccessHandle::FileSystemSyncAccessHandle(ScriptExecutionContext& c
+    if (m_isClosed)
+        return Exception { InvalidStateError, "AccessHandle is closed"_s };
 
     auto* scope = downcast<WorkerGlobalScope>(scriptExecutionContext());
     if (!scope)
-        return promise.reject(Exception { InvalidStateError, "Context is invalid"_s });
-
-    m_pendingPromises.append(WTFMove(promise));
-    WorkerGlobalScope::postFileSystemStorageTask([weakThis = WeakPtr { *this }, file = m_file.handle(), workerThread = Ref { scope->thread() }]() mutable {
-        workerThread->runLoop().postTask([weakThis = WTFMove(weakThis), success = FileSystem::flushFile(file)](auto&) mutable {
-            if (weakThis)
-                weakThis->completePromise(success ? ExceptionOr<void> { } : Exception { UnknownError });
-        });
-    });
-}
+        return Exception { InvalidStateError, "Context is invalid"_s };
 
-void FileSystemSyncAccessHandle::close(DOMPromiseDeferred<void>&& promise)
-{
-    closeInternal([weakThis = WeakPtr { *this }, promise = WTFMove(promise)](auto result) mutable {
-        promise.settle(WTFMove(result));
-    });
+    auto result = FileSystem::fileSize(m_file.handle());
+    return result ? ExceptionOr<unsigned long long> { result.value() } : Exception { UnknownError };
 }
 
-void FileSystemSyncAccessHandle::closeInternal(CloseCallback&& callback)
+ExceptionOr<void> FileSystemSyncAccessHandle::flush()
 {
-    if (m_closeResult)
-        return callback(ExceptionOr<void> { *m_closeResult });
+    if (m_isClosed)
+        return Exception { InvalidStateError, "AccessHandle is closed"_s };
 
-    auto isClosing = !m_closeCallbacks.isEmpty();
-    m_closeCallbacks.append(WTFMove(callback));
-    if (isClosing)
-        return;
+    auto* scope = downcast<WorkerGlobalScope>(scriptExecutionContext());
+    if (!scope)
+        return Exception { InvalidStateError, "Context is invalid"_s };
 
-    ASSERT(m_file);
-    closeFile();
+    bool succeeded = FileSystem::flushFile(m_file.handle());
+    return succeeded ? ExceptionOr<void> { } : Exception { UnknownError };
 }
 
-void FileSystemSyncAccessHandle::closeFile()
+ExceptionOr<void> FileSystemSyncAccessHandle::close()
 {
-    if (!m_file)
-        return;
+    if (m_isClosed)
+        return Exception { InvalidStateError, "AccessHandle is closed"_s };
 
     auto* scope = downcast<WorkerGlobalScope>(scriptExecutionContext());
-    ASSERT(scope);
-
-    WorkerGlobalScope::postFileSystemStorageTask([weakThis = WeakPtr { *this }, file = std::exchange(m_file, { }), workerThread = Ref { scope->thread() }]() mutable {
-        workerThread->runLoop().postTask([weakThis = WTFMove(weakThis)](auto&) mutable {
-            if (weakThis)
-                weakThis->didCloseFile();
-        });
-    });
-}
+    if (!scope)
+        return Exception { InvalidStateError, "Context is invalid"_s };
 
-void FileSystemSyncAccessHandle::didCloseFile()
-{
-    closeBackend(CloseMode::Async);
+    closeInternal(ShouldNotifyBackend::Yes);
```

Comment:
In case of ShouldNotifyBackend::Yes, should we block the worker thread until getting the notification from the backend?
If we do not block, could there be race conditions happening within the web app (say the web app has two workers doing IO?)

Reply:
Could you elaborate on the race condition case? A file can only have one FileSystemSyncAccessHandle at a time.

Reply:
I am thinking two web workers, the first one calls close and posts a message to the second one which immediately creates a sync handle on the same file entry.
A race condition between the close and the creation could theoretically happen although I am not sure this can happens today.
Say if the two web workers are in different processes and the postMessage does not go through the networking process.

Reply:
I see, so the theoretical case is 
1. worker1 closes access handle, sends message "close access handle" to network process and sends message "start" to worker2
2. worker2 receives the "start" message, and sends message "create access handle" to network process
3. network process receives "create access handle" message from worker 1 before "close access handle" from worker, and will fail the "create access handle" message

Is it possible that two web processes can talk to each other without going through network process? (The SyncAccessHandle is only available in DedicatedWorker now)

Reply:
In the past postMessage was through UIProcess, now it is going through NetworkProcess, though it might be handled in a different thread than file sync.
WebLock is another possibility and is UIProcess based IIRC.

Reply:
It feels like a rare case that network process will handle "create" message before the "close" message, as the "close access handle" message will be sent directly from web process 1 to network process, and the "create access handle" message from web process 2 will be sent after:
1. UI process receives the "release lock" message from web process 1
2. web process 2 receives "acquire lock" message from UI process

We could make the change if it turns out to be an issue I guess; blocking worker thread on close seems not too good for performance.

Reply:
Agreed this is rare, but this is the kind of issues that are very hard to diagnose.
Could you check what Chrome is doing here?

The spec is loose in the definition of close, it seems your interpretation might be correct since the only explicit step is to set the state. It might be worth having a more precise definition of close.

Reply:
It seems Chrome waits until close message is replied; I will make the change to be compatible (https://bugs.webkit.org/show_bug.cgi?id=250194).

Reply:
I filed https://github.com/whatwg/fs/issues/83 to keep track of this on spec side.